### PR TITLE
merge qmmm

### PR DESCRIPTION
**Pull request by MoleOrbitalHybridAnalyst**

### CODE REVIEW

File: gpu4pyscf/qmmm/pbc/itrf.py [hunk 1]
```diff
@@ -0,0 +1,1131 @@
+# Author: Chenghan Li
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/itrf.py:1]: Add the header similar to the other files. For example, https://github.com/pyscf/gpu4pyscf/blob/master/gpu4pyscf/pop/esp.py

File: gpu4pyscf/qmmm/pbc/itrf.py [hunk 2]
```diff
@@ -0,0 +1,1131 @@
...
+        method_class = scf_method.__class__
+
+    else:
+        if isinstance(scf_method._scf, _QMMM):
+            scf_method._scf.mm_mol = mm_mol
+            return scf_method
+
+        method_class = scf_method._scf.__class__
+
+    class QMMM(_QMMM, method_class):
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/itrf.py:116]: Please follow the latest design of the QMMM class in PySCF, which removed all the dynamic classes. https://github.com/pyscf/pyscf/blob/master/pyscf/qmmm/itrf.py#L113

File: gpu4pyscf/qmmm/pbc/itrf.py [hunk 3]
```diff
@@ -0,0 +1,1131 @@
...
+
+        def get_qm_ewald_pot(self, mol, dm, qm_ewald_hess=None):
+            # hess = d^2 E / dQ_i dQ_j, d^2 E / dQ_i dD_ja, d^2 E / dDia dDjb, d^2 E/ dQ_i dO_jab
+            if qm_ewald_hess is None:
+                qm_ewald_hess = self.mm_mol.get_ewald_pot(mol.atom_coords())
+                self.qm_ewald_hess = qm_ewald_hess
+            charges = self.get_qm_charges(dm)
+            dips = self.get_qm_dipoles(dm)
+            quads = self.get_qm_quadrupoles(dm)
+            ewpot0  = cp.einsum('ij,j->i', qm_ewald_hess[0], charges)
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/itrf.py:146]: Please use `contract` in cupy_helper instead of `cp.einsum'.

File: gpu4pyscf/qmmm/pbc/itrf.py [hunk 4]
```diff
@@ -0,0 +1,1131 @@
...
+            logger.note(self, '%d MM charges see directly QM density'%charges.shape[0])
+            if mm_mol.charge_model == 'gaussian' and len(coords) != 0:
+                expnts = cp.hstack([mm_mol.get_zetas()] * len(Ls))[mask]
+                # FIXME slice mm coords when memory not enough
+                fakemol = gto.fakemol_for_charges(coords.get(), expnts.get())
+
+                intopt = int3c2e.VHFOpt(mol, fakemol, 'int2e')
+                intopt.build(self.direct_scf_tol, diag_block_with_triu=True, aosym=True, 
+                             group_size=int3c2e.BLKSIZE, group_size_aux=int3c2e.BLKSIZE)
+
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/itrf.py:205]: These calculation has been implemented in `get_j_int3c2e_pass1`. It is more efficient and consumes less memory.  And we may further improve this later.
https://github.com/pyscf/gpu4pyscf/blob/master/gpu4pyscf/solvent/pcm.py#L394

File: gpu4pyscf/qmmm/pbc/itrf.py [hunk 5]
```diff
@@ -0,0 +1,1131 @@
...
+    # Avoid to initialize _QMMMGrad twice
+    if isinstance(scf_grad, _QMMMGrad):
+        return scf_grad
+
+    assert (isinstance(scf_grad.base, scf.hf.SCF) and
+           isinstance(scf_grad.base, _QMMM))
+
+    grad_class = scf_grad.__class__
+    class QMMM(_QMMMGrad, grad_class):
+        def dump_flags(self, verbose=None):
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/itrf.py:565]: dump_flags = pyscf.QMMM.dump_flags

File: gpu4pyscf/qmmm/pbc/itrf.py [hunk 6]
```diff
@@ -0,0 +1,1131 @@
...
+                Rij = R
+                Tij = 1 / r
+                # Tija   = \nabla_a Tij
+                # Tijab  = \nabla_a Tijb
+                # Tijabc = \nabla_a Tijbc
+                Tija = -cp.einsum('ijx,ij->ijx', Rij, Tij**3)
+                Tijab   = 3 * cp.einsum('ija,ijb->ijab', Rij, Rij) 
+                Tijab   = cp.einsum('ijab,ij->ijab', Tijab, Tij**5)
+                Tijab  -= cp.einsum('ij,ab->ijab', Tij**3, np.eye(3))
+                Tijabc  = -15 * cp.einsum('ija,ijb,ijc->ijabc', Rij, Rij, Rij)
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/itrf.py:698]: Three-tensor contractions in cp.einsum consumes a large amount of GPU memory. Please decompose them into two tensor contractions manually, and replace them with `contract` in cupy_helper.

File: gpu4pyscf/qmmm/pbc/itrf.py [hunk 7]
```diff
@@ -0,0 +1,1131 @@
...
+                # difference between MM gaussain charges and MM point charges
+                mask = dist2 > cell.rcut_hcore**2
+                zetas = cp.asarray(cell.get_zetas())
+                min_expnt = cp.min(zetas)
+                max_ewrcut = pbc.gto.cell._estimate_rcut(min_expnt, 0, 1., cell.precision)
+                cut2 = (max_ewrcut + rmax_qm)**2
+                mask = mask & (dist2 <= cut2)
+                expnts = cp.hstack([cp.sqrt(zetas)] * len(Lall))[mask]
+                Tija, Tijab, Tijabc = grad_kTij(R[:,mask], r[:,mask], expnts)
+                qm_ewovrl_grad[i0:i1] -= cp.einsum('i,ijx,j->ix', qm_charges[i0:i1], Tija, all_mm_charges[mask])
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/itrf.py:787]: The same tensor contraction issue here. `cp.einsum` behaves differently from `np.einsum`.

File: gpu4pyscf/qmmm/pbc/mm_mole.py [hunk 8]
```diff
@@ -0,0 +1,391 @@
+import numpy
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/mm_mole.py:1]: Add the header

File: gpu4pyscf/scf/hf.py [hunk 9]
```diff
@@ -700,7 +700,7 @@ class RHF(SCF):
     to_gpu = utils.to_gpu
     device = utils.device
 
-    _keys = {'e_disp', 'h1e', 's1e', 'e_mf', 'conv_tol_cpscf', 'disp_with_3body'}
+    _keys = {'e_disp', 'h1e', 's1e', 'e_mf', 'conv_tol_cpscf', 'disp_with_3body', 'h1_on_cpu'}
```
Comment (wxj6000) [inline gpu4pyscf/scf/hf.py:703]: Is `h1_on_cpu` necessary?

**Comment by wxj6000** [comment]
@MoleOrbitalHybridAnalyst Please add unit test and corresponding examples in the example folder.

**Comment by wxj6000** [comment]
Please refer to the contribution guideline. https://github.com/pyscf/gpu4pyscf/blob/master/CONTRIBUTING.md

File: gpu4pyscf/qmmm/pbc/mm_mole.py (hunk 8 above)
Comment (MoleOrbitalHybridAnalyst) [reply gpu4pyscf/qmmm/pbc/mm_mole.py:18]: added

File: gpu4pyscf/qmmm/pbc/itrf.py (hunk 4 above)
Comment (MoleOrbitalHybridAnalyst) [reply gpu4pyscf/qmmm/pbc/itrf.py:461]: herein I want to keep hcore w/o contracted with dm.

**Comment by MoleOrbitalHybridAnalyst** [comment]
all comments addressed

File: examples/23-qmmm_pbc.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import pyscf
+from pyscf import lib
+from gpu4pyscf.scf import hf
+from gpu4pyscf.qmmm.pbc import itrf, mm_mole
+
+import numpy as np
+
+lib.num_threads(2)
```
Comment (wxj6000) [inline examples/23-qmmm_pbc.py:23]: It is not necessary to set up the number of threads.

File: gpu4pyscf/qmmm/pbc/itrf.py (hunk 4 above)
Comment (wxj6000) [reply gpu4pyscf/qmmm/pbc/itrf.py:461]: You can contract the integrals with charges by `get_j_int3c2e_pass2`.

File: gpu4pyscf/qmmm/pbc/tests/test_qmmm.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+# Author: Chenghan Li <lch004218@gmail.com>
+
+import unittest
+import numpy as np
+import pyscf
+from pyscf import lib
+from gpu4pyscf.dft import rks
+from gpu4pyscf.qmmm.pbc import itrf
+
+lib.num_threads(2)
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/tests/test_qmmm.py:25]: This is not necessary.

File: gpu4pyscf/qmmm/pbc/tests/test_qmmm.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    mol.stdout.close()
+    del mol
+
+def run_dft(xc):
+    mf = rks.RKS(mol, xc=xc).density_fit(auxbasis=auxbasis)
+    mf = itrf.add_mm_charges(
+        mf, [[1,2,-1],[3,4,5]], np.eye(3)*15, [-5,5], [0.8,1.2], rcut_ewald=8, rcut_hcore=6)
+    mf.conv_tol = scf_tol
+    mf.max_cycle = max_scf_cycles
+    mf.screen_tol = screen_tol
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/tests/test_qmmm.py:58]: `screen_tol` has been deprecated. Use `direct_scf_tol` instead.

**Comment by MoleOrbitalHybridAnalyst** [comment]
@wxj6000 comments resolved

**Comment by wxj6000** [comment]
> @wxj6000 comments resolved

Please resolve the code style checks.

**Comment by MoleOrbitalHybridAnalyst** [comment]
> > @wxj6000 comments resolved
> 
> Please resolve the code style checks.

fixed

File: gpu4pyscf/qmmm/pbc/itrf.py (hunk 1 above)
Comment (wxj6000) [reply gpu4pyscf/qmmm/pbc/itrf.py]: @MoleOrbitalHybridAnalyst The header is missing.

File: gpu4pyscf/qmmm/pbc/mm_mole.py (hunk 8 above)
Comment (wxj6000) [reply gpu4pyscf/qmmm/pbc/mm_mole.py:18]: It is missing again.

File: gpu4pyscf/qmmm/pbc/tools.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    mem_avail = cupy_helper.get_avail_mem()
+    blksize = int(mem_avail/64/3/(1+len(coords2)))
+    if blksize == 0:
+        raise RuntimeError(f"Not enough GPU memory, mem_avail = {mem_avail}, blkszie = {blksize}")
+    ene = 0
+    for i0, i1 in lib.prange(0, len(coords1), blksize):
+        Rij = coords1[i0:i1,None,:] - coords2[None,:,:]
+        rij = cp.linalg.norm(Rij, axis=-1)
+        Tij = 1 / rij
+        Tijabc  = -15 * cp.einsum('ija,ijb,ijc->ijabc', Rij, Rij, Rij)
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/tools.py:54]: `cupy.einsum` is still used here.

File: gpu4pyscf/qmmm/pbc/mm_mole.py [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+from cupyx.scipy.special import erf, erfc
+from pyscf import lib
+
+import cupy as cp
+from gpu4pyscf.df.df import ALIGNED, MIN_BLK_SIZE
+from gpu4pyscf.lib import cupy_helper
+
+contract = cupy_helper.contract
+
+class Cell(qmmm.mm_mole.Mole, pbc.gto.Cell):
```
Comment (wxj6000) [inline gpu4pyscf/qmmm/pbc/mm_mole.py:35]: This class should be available in PySCF too.

**Comment by MoleOrbitalHybridAnalyst** [comment]
@wxj6000 resolved all

**Comment by MoleOrbitalHybridAnalyst** [comment]
@wxj6000 removing h1_on_cpu will result in my hcore running in the background with our GPU tasks. This occasionally gives wrong integrals (although I don't know the exact reasons). So I want my hcore not running in the background. Is it possible without changing scf.hf etc. ?

**Pull request merged**